### PR TITLE
Fix cmake export for grpc

### PR DESCRIPTION
**Pull request by hacst**
This patch fixes two issues with the targets exported by the
grpc cmake build:

- Install directories are now given relative instead of absolute.
  Previously the exported targets would contain absolute paths to
  their created targets when they should be relative. The actual
  installation location stays the same as cmake automatically
  interprets the paths relative to CMAKE_INSTALL_PREFIX

- ZLIB and OpenSSL are now linked using targets instead of the
  legacy library variables if possible. Using the legacy paths
  produces absolute paths to those dependencies in the export.
  With this change cmake inserts the target names instead.

With these changes no more absolute paths are inserted into the
export allowing the install location to be used as an artifact
that can be moved or transfered to another machine and used there.

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @hacst isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by hacst** [comment]
I signed it!

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by jtattermusch** [comment]
Looks like this breaks our windows distribution test (install prerequisites, install grpc, build helloworld example):
Here's the log:
https://source.cloud.google.com/results/invocations/b8de957b-8d73-447c-9b52-b32dde1cceec/log

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by hacst** [comment]
I am not quite sure what is happening here. It looks like the build somehow mixes msys standard library headers with visual studio cl.exe and the normal Visual Studio headers.

> C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\CL.exe /c /IT:\src\github\grpc\workspace_cpp_windows_x86_cmake\testinstall\include /IT:\src\github\grpc\workspace_cpp_windows_x86_cmake\examples\cpp\helloworld\cmake\build **/IC:\tools\msys64\usr\include** /nologo /W3 /WX- /O2 /Ob2 /Oy- /D WIN32 /D _WINDOWS /D NDEBUG /D _WIN32_WINNT=0x600 /D "CMAKE_INTDIR=\"Release\"" /D _MBCS /Gm- /EHsc /MD /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"greeter_async_client.dir\Release\\" /Fd"greeter_async_client.dir\Release\vc140.pdb" /Gd /TP /analyze- /errorReport:queue T:\src\github\grpc\workspace_cpp_windows_x86_cmake\examples\cpp\helloworld\greeter_async_client.cc T:\src\github\grpc\workspace_cpp_windows_x86_cmake\examples\cpp\helloworld\cmake\build\helloworld.pb.cc T:\src\github\grpc\workspace_cpp_windows_x86_cmake\examples\cpp\helloworld\cmake\build\helloworld.grpc.pb.cc
> 
> 
>   greeter_async_client.cc
> 
> 
> **C:\tools\msys64\usr\include\sys/cdefs.h**(704): warning C4005: '__guarded_by': macro redefinition [T:\src\github\grpc\workspace_cpp_windows_x86_cmake\examples\cpp\helloworld\cmake\build\greeter_async_client.vcxproj]
> 
> 
>   **C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include**\ConcurrencySal.h(317): note: see previous definition of '__guarded_by'

That of course wreaks complete havoc over the the build. It does stuff like taking math.h from msys (which is not compatible with cl.exe to begin with) and then cmath from VS which is just a bad time.

Is there more information available than the log or a way for me to reproduce this test environment locally? At this point I am quite confused about how my changes would cause an msys include path to be added.

**Comment by hacst** [comment]
Ok. This was a bit painful to figure out. Turns out the problem was in ```run_distrib_test_cmake.bat``` . The cmake call there was incomplete making cmake find a wrong zlib include directory from the environment now that the paths are no longer hardcoded in the grpc export. Explicitely specifying zlib locations like in the other cmake calls in that file should fix the build issues.

Here the commit message with a more detailed explanation:

> Fix helloworld zlib include directory confusion in run_distrib_test_cmake.bat
> 
> We have to give cmake the explicit zlib location as the internal testing
> environment has C:/msys64/usr/bin in the PATH and zlib-devel installed.
> cmake has heuristics in find_package which sees this /bin suffix in PATH
> and adds C:/msys64/usr/ to the find_package search locations. Doing a
> find_package(zlib) in this environment makes the find module for it (FindZLIB)
> find the header zlib.h in C:/msys64/usr/include while the library will still
> be taken from the testinstall location masking the issue in the log. To satisfy
> the dependency cmake adds C:/msys64/usr/include as an include directory. This
> makes cl.exe build with mixed C and C++ standard lib headers breaking the
> build.
> 
> This issue was previously masked by cmake writing absolute paths for zlib and
> other dependencies into the grpc cmake exports.

Regardless of this specific issue: I'm not sure it is a good idea to have msys in your PATH when using cmake. cmake is just so aggressive when trying to find dependencies that it feels like asking for trouble.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: test/distrib/cpp/run_distrib_test_cmake.bat [hunk 1]
```diff
@@ -64,7 +64,7 @@ mkdir cmake
 cd cmake
 mkdir build
 cd build
-cmake -DCMAKE_INSTALL_PREFIX=%INSTALL_DIR% -DOPENSSL_ROOT_DIR=%OPENSSL_DIR% -DOPENSSL_INCLUDE_DIR=%OPENSSL_DIR%/include ../.. || goto :error
+cmake -DCMAKE_INSTALL_PREFIX=%INSTALL_DIR% -DOPENSSL_ROOT_DIR=%OPENSSL_DIR% -DOPENSSL_INCLUDE_DIR=%OPENSSL_DIR%/include -DZLIB_LIBRARY=%INSTALL_DIR%/lib/zlibstatic.lib -DZLIB_INCLUDE_DIR=%INSTALL_DIR%/include ../.. || goto :error
```
Comment (jtattermusch) [inline test/distrib/cpp/run_distrib_test_cmake.bat]: Nice find, thanks for the detailed description in the commit message. 
I realized I'm still unhappy with the way zlib gets included - ideally setting ZLIB_ROOT should be enough here.
(as per https://cmake.org/cmake/help/v3.0/module/FindZLIB.html, where only ZLIB_ROOT is an allowed hint).
I'll see what the tests say.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: test/distrib/cpp/run_distrib_test_cmake.bat (hunk 1 above)
Comment (jtattermusch) [reply test/distrib/cpp/run_distrib_test_cmake.bat]: I added a commit with some cleanup and it seems that the tests are passing.

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/14334
https://github.com/grpc/grpc/issues/13848

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_pump.BM_PumpStreamServerToClient_SockPair__0.counters.new: 1


[microbenchmarks] No significant performance differences
```

**Comment by jtattermusch** [comment]
@hacst  Thanks for contributing!

**Comment by hacst** [comment]
@jtattermusch Awesome. Thanks.

**Pull request merged**